### PR TITLE
[fix]: [CCM-8935]: Separate includeOthers data with BMOthers data in chart

### PR DESCRIPTION
**Pull request by iamanmolm**
## Describe your changes
This fix basically includes where we don't allow user to have unallocated cost bucket name as either "Others" or "Unallocated" in the cost category.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39385)

### CODE REVIEW

File: 450-ce-views/src/main/java/io/harness/ccm/views/businessMapping/service/impl/BusinessMappingServiceImpl.java [hunk 1]
```diff
@@ -121,6 +124,17 @@ private void validateBusinessMapping(final BusinessMapping businessMapping) {
     updateBusinessMapping(businessMapping);
   }
 
+  @Override
+  public boolean isValidBusinessMappingUnallocatedCostLabel(final BusinessMapping businessMapping) {
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/businessMapping/service/impl/BusinessMappingServiceImpl.java:128]: Here, we can directly return the response.
```
@Override
  public boolean isValidBusinessMappingUnallocatedCostLabel(final BusinessMapping businessMapping) {
    return !(businessMapping != null && businessMapping.getUnallocatedCost() != null
        && businessMapping.getUnallocatedCost().getLabel() != null
        && (businessMapping.getUnallocatedCost().getLabel().equalsIgnoreCase(OTHERS)
            || businessMapping.getUnallocatedCost().getLabel().equalsIgnoreCase(UNALLOCATED)));
  }
```

Comment (iamanmolm) [reply 450-ce-views/src/main/java/io/harness/ccm/views/businessMapping/service/impl/BusinessMappingServiceImpl.java:128]: Done

**Review by sdsingh07 — DISMISSED** [review]
:lgtm:

> Reviewed 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39385#-:-NFr191O0bHzTsPQSEjT:bnfp4nl)* status: :shipit: complete! ✓

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java [hunk 2]
```diff
@@ -71,6 +73,13 @@ public RestResponse<BusinessMapping> save(
     if (!businessMappingService.isNamePresent(businessMapping.getName(), businessMapping.getAccountId())) {
       throw new InvalidRequestException("Cost category name already exists.");
     }
+    if (businessMapping != null && businessMapping.getUnallocatedCost() != null
```
[the remark this answers is not in the document]
Comment (iamanmolm) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java]: Made a common function. Confirmed with Rohit, we are doing it only for exact matches.

[the remark this answers is not in the document]
Comment (sdsingh07) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java]: Then, we have to remove the `equalsIgnoreCase` check.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java [hunk 3]
```diff
@@ -71,6 +71,9 @@ public RestResponse<BusinessMapping> save(
     if (!businessMappingService.isNamePresent(businessMapping.getName(), businessMapping.getAccountId())) {
       throw new InvalidRequestException("Cost category name already exists.");
     }
+    if (businessMappingService.isValidBusinessMappingUnallocatedCostLabel(businessMapping)) {
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:74]: If it's not valid, then throw exception.
`if (!businessMappingService.isValidBusinessMappingUnallocatedCostLabel(businessMapping))`

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java [hunk 4]
```diff
@@ -112,6 +115,9 @@ public RestResponse<String> update(
         throw new InvalidRequestException("Cost category name already exists.");
       }
     }
+    if (businessMappingService.isValidBusinessMappingUnallocatedCostLabel(businessMapping)) {
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:118]: If it's not valid, then throw exception.
`if (!businessMappingService.isValidBusinessMappingUnallocatedCostLabel(businessMapping))`

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java (hunk 3 above)
Comment (iamanmolm) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java]: Already resolved.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java (hunk 4 above)
Comment (iamanmolm) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java]: Already resolved.

**Review by sdsingh07 — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39385#-:-NFr7Wx5CLiUT1PAHSzj:bnfp4nl)* status: :shipit: complete! ✓

**Comment by iamanmolm** [comment]
trigger compile, ft1, ti0, ti1, ti2, ti3

**Review by sdsingh07 — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39385)* status: :shipit: complete! ✓

**Pull request merged**